### PR TITLE
SLM-13 expired barcode service

### PR DESCRIPTION
**Pull request by mikehalmamoj**
* Adds a new config parameter to define barcode expiry duration
* Adds a new service to check for an expired barcode
* Refactor existing config property classes to live in same package as used
* Refactor some unit tests to remove noise

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeEventService.kt [hunk 1]
```diff
@@ -37,4 +42,20 @@ class BarcodeEventService(private val barcodeEventRepository: BarcodeEventReposi
         )
         throw ValidationException(Duplicate(firstCheck.createdDateTime, firstCheck.location))
       }
+
+  fun checkForExpired(barcode: Barcode, userId: String, location: String) =
+    barcodeEventRepository.findByBarcodeAndStatusOrderByCreatedDateTime(barcode, BarcodeStatus.CREATED)
+      .firstOrNull()
+      ?.takeIf { createdEvent -> createdEvent.createdDateTime < Instant.now().minus(barcodeConfig.expiry) }
```
Comment (nathanrussell-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeEventService.kt:49]: Is there a (remote) edge case here that the barcode expiry could be xDays, then a barcode is created on day 1 (and hence its expiry is 1+x), but before it is scanned and checked (ie. its in the physical postal system) the barcode expiry config is updated to yDays. Therefore when the barcode is scanned we dont know its real expiry date - all we know is its creation date and the current expiry config.

I know its a real edge case, but I don't know if its an issue? If we think its a issue I guess it means a new column on the Barcode table of its expiry date, that is computed at the point of creation ?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/config/ExceptionHandler.kt [hunk 2]
```diff
@@ -103,10 +104,11 @@ object EmailInvalid : MagicLinkRequestErrorCodes("INVALID_EMAIL", "Enter an emai
 object EmailInvalidCjsm : MagicLinkRequestErrorCodes("INVALID_CJSM_EMAIL", "Enter an email address which ends with 'cjsm.net'")
 
 sealed class CheckBarcodeErrorCodes(code: String, userMessage: String) : StandardErrorCodes(code, userMessage)
-class Duplicate(val scannedDate: Instant, val scannedLocation: String) : CheckBarcodeErrorCodes("DUPLICATE", "Someone scanned this barcode ${scannedDate.format()} at $scannedLocation. It may be an illegal copy.")
+class Duplicate(val scannedDate: Instant, val scannedLocation: String) : CheckBarcodeErrorCodes("DUPLICATE", "Someone scanned this barcode ${scannedDate.formatAtTimeOnDate()} at $scannedLocation. It may be an illegal copy.")
+class Expired(val createdDate: Instant, val barcodeExpiry: Duration) : CheckBarcodeErrorCodes("EXPIRED", "This barcode was created ${createdDate.ageInDays()}, ${createdDate.formatOnDate()}.")
```
Comment (nathanrussell-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/config/ExceptionHandler.kt:108]: Would it be better for this Expired class to have the actual expiry date, rather than an expiry duration? As it stands it means consumers of the service (the front end express app) have to work out the expiry date

File: src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/config/ExceptionHandler.kt [hunk 3]
```diff
@@ -103,10 +104,11 @@ object EmailInvalid : MagicLinkRequestErrorCodes("INVALID_EMAIL", "Enter an emai
...
 
 private val timeFormatter = DateTimeFormatter.ofPattern("h:mm a").withZone(ZoneId.systemDefault())
 private val dateFormatter = DateTimeFormatter.ofPattern("d MMMM y").withZone(ZoneId.systemDefault())
-private fun Instant.format(): String {
-  return """at ${timeFormatter.format(this).lowercase()} on ${dateFormatter.format(this)}"""
-}
+private fun Instant.formatAtTimeOnDate() = """at ${timeFormatter.format(this).lowercase()} ${this.formatOnDate()}"""
```
Comment (nathanrussell-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/config/ExceptionHandler.kt:112]: I love kotlin extension functions 😄 👍

File: src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeEventServiceTest.kt [hunk 4]
```diff
@@ -8,34 +8,36 @@ import com.nhaarman.mockitokotlin2.verify
...
 import java.time.Instant
 import java.time.temporal.ChronoUnit
 import javax.persistence.EntityNotFoundException
 
 class BarcodeEventServiceTest {
 
   private val barcodeEventRepository = mock<BarcodeEventRepository>()
-  private val barcodeEventService = BarcodeEventService(barcodeEventRepository)
+  private val barcodeConfig = mock<BarcodeConfig>()
+  private val barcodeEventService = BarcodeEventService(barcodeEventRepository, barcodeConfig)
```
Comment (nathanrussell-moj-digital) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeEventServiceTest.kt:29]: Not that I have any strong feelings about this (other than how I've always done things in the past), but do we not use the mockito annotations? Like this:
```
@ExtendWith(MockitoExtension::class)
class BarcodeEventServiceTest {

  @Mock
  private lateinit var barcodeEventRepository: BarcodeEventRepository

  @Mock
  private lateinit var barcodeConfig: BarcodeConfig

  @InjectMocks
  private lateinit var barcodeEventService: BarcodeEventService
```

File: src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeEventServiceTest.kt [hunk 5]
```diff
@@ -44,18 +46,18 @@ class BarcodeEventServiceTest {
   inner class CheckForCreated {
     @Test
     fun `should do nothing if created event exists`() {
-      whenever(barcodeEventRepository.findByBarcodeAndStatusOrderByCreatedDateTime(any(), eq(BarcodeStatus.CREATED)))
-        .thenReturn(listOf(BarcodeEvent(1L, Barcode("SOME_BARCODE"), "any_user", BarcodeStatus.CREATED)))
+      mockFindBarcodeEvents(CREATED, listOf(aBarcodeEvent(status = CREATED)))
 
-      barcodeEventService.checkForCreated(Barcode("SOME_BARCODE"))
+      assertDoesNotThrow {
```
Comment (nathanrussell-moj-digital) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeEventServiceTest.kt:51]: 👍

File: src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeServiceTest.kt [hunk 6]
```diff
@@ -59,33 +63,42 @@ class BarcodeServiceTest {
...
         .isEqualTo(expectedException)
 
       verify(barcodeEventService)
-        .createEvent(Barcode("SOME_BARCODE"), "current_user", BarcodeStatus.CHECKED, "current_location")
+        .createEvent(aBarcode(), "current_user", BarcodeStatus.CHECKED, "current_location")
     }
   }
+
+  // Test helpers
+  private fun aBarcode() = Barcode("SOME_BARCODE")
```
Comment (nathanrussell-moj-digital) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeServiceTest.kt:97]: Given that this method is identical to the one in BarcodeEventServiceTest, would it be worth factoring this (and aBarcodeEvent from BarcodeEventServiceTest) into something that can be re-used by other tests?

In previous projects we had a concept of test data factories. They weren't 'factories' in the strictest sense of the design pattern, but basically were classes with public static methods that effectively did what you are doing here. (they were java projects rather that kotlin, hence static methods, but you get the idea)
Typically we'd have a package `testsupport` or similar, and we'd have subpackages for test data factories like this, and packages for other test support gubbins.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/config/ExceptionHandlerTest.kt [hunk 7]
```diff
@@ -13,4 +17,15 @@ class ExceptionHandlerTest {
     assertThat(ex.code).isEqualTo("DUPLICATE")
     assertThat(ex.userMessage).contains("at 9:11 am on 8 December 2021")
   }
+
+  @Test
```
Comment (nathanrussell-moj-digital) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/config/ExceptionHandlerTest.kt:21]: Happy enough to keep this test, but I am questioning what its actually testing. I dont think its testing creating the exception as the test name suggests - I think its testing the internals of the exception class itself ?
Given the test class is ExceptionHandlerTest, and this test does nothing with the ExceptionHandler itself, I wonder if its in the wrong place?

**Review by nathanrussell-moj-digital — APPROVED** [review]
Discussed comments with Mike just now - happy with agreed outcomes ✅

File: src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeEventService.kt (hunk 1 above)
Comment (mikehalmamoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeEventService.kt:49]: It's a really good point... though it is very edge case. And also very hard to predict what the requirement would be if the expiry duration changed - they might be happy to apply the new rule retrospectively anyway. BUT if we go with your suggestion then at least we can be flexible either way.

I'm going to have to sleep on this one...

File: src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeEventServiceTest.kt (hunk 4 above)
Comment (mikehalmamoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeEventServiceTest.kt:29]: It's become a convention in HMPPS not to use InjectMocks... I think because there are some edge cases where it doesn't work very well (maybe something to do with constructor defaults?) but tbh I can't remember the detail.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeServiceTest.kt (hunk 6 above)
Comment (mikehalmamoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeServiceTest.kt:97]: As discussed I'm not keen on test factories due to the levels of indirection it introduces, but where there is a need I'd rather keep them in the same test class to minimise indirection. That said, if we find ourselves duplicating these helpers in more places then we should review and consider the benefits of a testsupport package.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/config/ExceptionHandlerTest.kt (hunk 7 above)
Comment (mikehalmamoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/config/ExceptionHandlerTest.kt:21]: As discussed, the underlying problem here is that the error codes are all hanging off the bottom of the ExceptionHandler... but they have become complex enough to warrant their own source file now. So yes I will move the error codes and rename these tests in my next PR.
